### PR TITLE
Add preliminary beaconing support

### PR DESCRIPTION
**Pull request by noraabiakar**
# PR for issue #1906

## What is being addressed

- Support parsing and writing json router configurations with beaconing settings.
- Respond to `timesync` messages from basics station with corresponding `timesync` message including GPS time. 

## How is this addressed

- `LnsStationConfiguration.RouterConfigurationConverter` works with optional `bcning` settings in the device twins. 
- `LnsProtocolMessageProcessor.HandleDataMessageAsync` responds to messages of type `LnsMessageType.TimeSync`.

### CODE REVIEW

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/Beaconing.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+        public Beaconing(uint dR, uint[] layout, uint[] freqs)
+        {
+            DR = dR;
+            this.layout = layout;
+            this.freqs = freqs;
+        }
+
+        public uint DR { get; set; }
+        public uint[] layout { get; set; }
```
Comment (fbeltrao) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/Beaconing.cs:17]: Should property names start with capital letter?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/TimeSyncMessage.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan.NetworkServer.BasicsStation.JsonHandlers
+{
+
+    using LoRaWan.NetworkServer.BasicsStation;
+
+    internal class TimeSyncMessage
+    {
+        public ulong txtime { get; set; }
```
Comment (fbeltrao) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/TimeSyncMessage.cs:11]: Should property names start with capital letter?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs [hunk 3]
```diff
@@ -196,9 +196,16 @@ internal async Task InternalHandleDataAsync(RouteValueDictionary routeValues, We
                     break;
                 case var messageType and (LnsMessageType.DownlinkMessage or LnsMessageType.RouterConfig):
                     throw new NotSupportedException($"'{messageType}' is not a valid message type for this endpoint and is only valid for 'downstream' messages.");
+                case LnsMessageType.TimeSync:
+                    var timeSyncData = JsonSerializer.Deserialize<TimeSyncMessage>(json);
+                    LogReceivedMessage(this.logger, "TimeSync", json, null);
+                    timeSyncData.gpstime = (ulong)DateTime.UtcNow.Subtract(
+                        new DateTime(1980, 1, 6, 0, 0, 0, DateTimeKind.Utc) // GPS Epoch
```
Comment (fbeltrao) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs:203]: Suggest making the GpsEpoch value a static read-only property, so we don't create it multiple times.

Comment (Mandur) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs:203]: Could we make  constant for this value, we might reuse it in the future?

File: Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs [hunk 4]
```diff
@@ -422,11 +555,12 @@ public void WriteRouterConfig_Throws_WhenInvalidSx1301Conf(string sx1301Conf)
                 ""nocca"": {7},
                 ""nodc"": {8},
                 ""nodwell"": {9}
+                {10}
             }}";
 
             static string Serialize(object obj) => JsonSerializer.Serialize(obj);
 
-            return string.Format(CultureInfo.InvariantCulture, Regex.Replace(template, "\\s+", string.Empty),
+           return string.Format(CultureInfo.InvariantCulture, Regex.Replace(template, "\\s+", string.Empty),
```
Comment (Mandur) [inline Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs:563]: This looks like of a formatting error

File: Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs [hunk 5]
```diff
@@ -256,6 +256,125 @@ public void WriteRouterConfig()
...
+                                         (SF10, BW125, false),
+                                         (SF9 , BW125, false),
+                                         (SF8 , BW125, false),
+                                         (SF7 , BW125, false),
+                                         (SF7 , BW250, false),
+                                     },
+                                     includeBcning: true,
+                                     flags: RouterConfigStationFlags.NoClearChannelAssessment | RouterConfigStationFlags.NoDutyCycle | RouterConfigStationFlags.NoDwellTimeLimitations) ;
+
+        var ValidRouterConfigMessageWithBcning = JsonUtil.Strictify(@"{
```
Comment (fbeltrao) [inline Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs:280]: Nit: adding /*lang=json*/ helps the editor understand the content of the text. It is a Visual Studio suggestion :)

```suggestion
        var ValidRouterConfigMessageWithBcning = JsonUtil.Strictify(/*lang=json*/ @"{
```

**Review by Mandur — APPROVED** [review]

File: Tests/Unit/NetworkServer/LnsProtocolMessageProcessorTests.cs [hunk 6]
```diff
@@ -195,6 +195,38 @@ public async Task InternalHandleDataAsync_ShouldSendExpectedJsonResponseType_For
...
+                               sentEnd = end;
+                           });
+
+            // act
+            await this.lnsMessageProcessorMock.InternalHandleDataAsync(this.httpContextMock.Object.Request.RouteValues,
+                                                                       this.socketMock.Object,
+                                                                       CancellationToken.None);
+
+            // assert
+            Assert.Contains(expectedSubstring, sentString, StringComparison.Ordinal);
```
Comment (fbeltrao) [inline Tests/Unit/NetworkServer/LnsProtocolMessageProcessorTests.cs:226]: Should we enhance the tests to check for:

- Response is a valid JSON
- The input `txttime` is available in the response
- The response has `gpstime` property containing a valid value (probably within a range).

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/LnsStationConfiguration.cs [hunk 7]
```diff
@@ -160,39 +160,57 @@ private class Sx1301Config
...
-                              JsonReader.Property("hwspec", JsonReader.String()),
-                              JsonReader.Property("freq_range", from r in JsonReader.Array(JsonReader.UInt32())
-                                                                select (new Hertz(r[0]), new Hertz(r[1]))),
-                              JsonReader.Property("DRs",
-                                  JsonReader.Array(from e in JsonReader.Tuple(JsonReader.Either(from n in JsonReader.UInt32().Validate(n => n == 0)
-                                                                                                select FskSpreadingFactor,
-                                                                                                JsonReader.UInt32().AsEnum(n => unchecked((SpreadingFactor)n))),
+        private static readonly IJsonReader<string> RouterConfigurationConverter = convertRouterConfiguration();
+
+        private static IJsonReader<string> convertRouterConfiguration()
```
Comment (fbeltrao) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/LnsStationConfiguration.cs:165]: Nit: method names should start with capital letter.

**Review by fbeltrao — APPROVED** [review]

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/LnsStationConfiguration.cs [hunk 8]
```diff
@@ -160,39 +160,57 @@ private class Sx1301Config
...
+                                 select (SpreadingFactor: e.Item1, Bandwidth: e.Item2, DownloadOnly: e.Item3)
+                                 into e
+                                 select e.SpreadingFactor is FskSpreadingFactor ? (FskSpreadingFactor, 0, e.DownloadOnly)
+                                     : Enum.IsDefined(e.Bandwidth) ? (e.SpreadingFactor, e.Bandwidth, e.DownloadOnly)
+                                     : throw new JsonException($"Invalid bandwidth: {e.Bandwidth}")));
+            var sx1301Reader = JsonReader.Property("sx1301_conf", JsonReader.Array(Sx1301ConfReader));
+            var noccaReader = JsonReader.Property("nocca", JsonReader.Boolean());
+            var nodcReader = JsonReader.Property("nodc", JsonReader.Boolean());
+            var nodwellReader = JsonReader.Property("nodwell", JsonReader.Boolean());
+            var bcningReader = JsonReader.Property("bcning",
```
Comment (danigian) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/LnsStationConfiguration.cs:197]: What if "bcning" is not there in the Twin? Isn't this going to fail as it is expecting it to be there?
Shouldn't we use the JsonReader.Either?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/LnsStationConfiguration.cs [hunk 9]
```diff
@@ -160,39 +160,57 @@ private class Sx1301Config
...
+                    JsonReader.Object(
+                        JsonReader.Property("DR", JsonReader.UInt32()),
+                        JsonReader.Property("layout", JsonReader.Array(JsonReader.UInt32())),
+                        JsonReader.Property("freqs", JsonReader.Array(JsonReader.UInt32())),
+                        (dRs, layout, freqs) => new Beaconing(dRs, layout, freqs)));
+
+            return JsonReader.Either(
```
Comment (danigian) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/LnsStationConfiguration.cs:204]: As discussed offline, let's use the default from Jsoneader.Property for bcningReader 👍 
Thanks for the efforts 😄

**Review by danigian — APPROVED** [review]

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/LnsStationConfiguration.cs (hunk 9 above)
Comment (noraabiakar) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/JsonHandlers/LnsStationConfiguration.cs]: Done! Thanks for the help :)

**Pull request merged**